### ISSUE TITLE
Battle test with JSX PPX V4

### ISSUE THREAD

**Issue by mattdamon108** [rescript-lang/syntax#615]
- [x] 1. You forgot to handle a possible case here, for example:
```rescript
// original
@react.component
let make = (~height, ~width, ~fill, ~className=?) => body

// error message
You forgot to handle a possible case here, for example: 
  {height: _, width: _, fill: Some(_), className: None, _}
| {height: _, width: _, fill: None, className: Some(_), _}
```
- [x] 2. The type variable name '_disabled is not allowed in programs
```rescript
// original
@react.component
let make = (~_disabled=?, ()) => body
```
- [x] 3. A type parameter occurs several times
- [x] 4. Values do not match between intf and impl
```
let make: props<option<string>, 'a, 'b, 'c> => React.element
is not included in
let make: React.componentLike<
props<string, string, int, React.element>
```

- [x] 5. At most one component definition per module or one gets an error of type `props` being defined multiple times "Multiple definition of the type name props".
- [x] 6. Explicit type annotation for optional arg in definition
This issue seems related to 4
```rescript
let make = (~x: option<int>=?) => body
// generated
type props<'x> = {key?: string, x?: 'x}
let make = ({?x, _}: props<option<int>>) => body
// error type mismatch option<int> vs. option<option<int>>
```
- [x] 7. Conversion for the alias
```rescript
let make = (~x as y) => body
```
- [x] 8. Backward compatibility for passing the optional expression to prop
This expression seems not working with V4, should work in different expression.
```rescript
x=?{a->returnOption} ❌ 

x=?{
  let _ = ()
  a->returnOption
} ⭕ 
```
- [x] 9. Raise error with loc

- [x] 10. If a component is called`foo` not `make` the ppx tries to call `make`.

- [x] 11. Error without location when a required prop is not supplied.

**Comment by mattdamon108** [comment]
WIP: #614

**Comment by mattdamon108** [comment]
I've looked into each issue and finally make my company project build successfully.
I added some new issues and work left to do though, but it won't take a long time.

**Comment by mattdamon108** [comment]
@cristianoc I have a question about the issue no. 8. I've investigated some and found this. I think it is a breaking change for users to have to change their code.
```rescript
type x = {x?: int}
let x = Some(1)

let a = {x: ?x}
let b = {x: ?x->Option.map(x => x)} // error
//           ^^^^^^^^^^^^^^^^^^^^ This has type: option<int>
//                                Somewhere wanted: int
let c = {x: ?Option.map(x, x => x)} // ok
```
I guess it might be related to the operator precedence or just the wrong syntax usage by users. As V4 parses the props and value into the record, it causes a build error there. If users have their code base expressing `let b` way, then it causes a breaking change. Can we make a workaround to prevent a break?

**Comment by cristianoc** [comment]
Can you add examples with and without parens? Then we know what's going on.

Examples checked in as tests. So we can look at the generated code.

**Comment by cristianoc** [comment]
How can that be a breaking change if the syntax with `?` did not exist before?

**Comment by mattdamon108** [comment]
> How can that be a breaking change if the syntax with `?` did not exist before?

Ah! I meant that breaking between V3 -> V4. I'll add the example in tests right a way.

**Comment by cristianoc** [comment]
One more:

```//test.res
  @react.component
  let anotherComponent = (~b) => {
    React.string(b)
  }
```

```
  We've found a bug for you!
  /Users/cristianocalcagno/GitHub/rescript-compiler/jscomp/gentype_tests/typescript-react-example/src/Hooks.res

  The value make can't be found
```

**Comment by mattdamon108** [comment]
https://github.com/rescript-lang/syntax/pull/617/commits/ead48312e5367b2c9ceb90755805fa224dc7ec22 The example added

**Comment by mattdamon108** [comment]
> ```rescript
>  @react.component
>   let anotherComponent = (~b) => {
>     React.string(b)
>   }
> ```

It is correct. As long as the parser do parse the uppercase jsx to `Foo.make`, the `make` is only allowed for the react component. I think the error message should be polished.

**Comment by cristianoc** [comment]
> ```rescript
> ```rescript
>  @react.component
>   let anotherComponent = (~b) => {
>     React.string(b)
>   }
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> 
> It is correct. As long as the parser do parse the uppercase jsx to Foo.make, the make is only allowed for the react component. I think the error message should be polished.
> EDIT: Plus, V4 allows only one react component per module.
> ```

The component does not need to be called `make`. This would be a very big breaking change.

**Comment by mattdamon108** [comment]
```rescript
@react.component
let make = (~b) => React.string(b)
```
It would be okay. Not sure I understand your point.

> The component does not need to be called make. 

Can you elaborate little more?

**Comment by cristianoc** [comment]
As in V3, the component can have any name. It can be called "default" to use file-level conventions for access from JS.
It could be called any name.

A direct `<Foo ....>` call, which obviously only works on components called `make`, is not the only way to consume a component.

**Comment by mattdamon108** [comment]
I finally got your point, I verified the same error. It seems an unreasonable error even without actually calling the component from outside. I'll look at it.

**Comment by mattdamon108** [comment]
I found the bug, I'll fix it soon!

**Comment by mattdamon108** [comment]
> As in V3, the component can have any name. It can be called "default" to use file-level conventions for access from JS. It could be called any name.
> 
> A direct `<Foo ....>` call, which obviously only works on components called `make`, is not the only way to consume a component.

This should be fixed https://github.com/rescript-lang/syntax/pull/617/commits/42032a2b599489e8dbdb2240f0291ca881885ee5

**Comment by cristianoc** [comment]
@mattdamon108 component definitions seems to support uncurried function definitions of `make` and treat them like curried ones. (In V3, this is an error). Is this intentional?

Asking as it would be interesting to figure out whether uncurried components could be supported in future. (Where the component application calls `make` an an uncurried function).

**Comment by mattdamon108** [comment]
That's not intentional, it is not in consideration actually.
It is very interesting. As far as I remember that the uncurried function will be the default, what would it affect the react component definition in the future? I expect that V3 will have a direct impact on the change, not V4. (V3 keep using labeled args as the original code, but V4 transforms labeled args to record as single arg)

**Comment by cristianoc** [comment]
I think that's the answer. Since V4 only ever takes 1 argument, it does not really matter if curried or uncurried is used. One can switch between the two easily.

**Comment by cristianoc** [comment]
One more issue.
An error without location is reported when a required prop is not supplied.

Example:
```rescript
module C = {
  @react.component
  let make = ( ~x, ~y) => React.string(x++y)
}

let c = <C x="abc" />
```

Compilation error:
```sh
FAILED: src/V4.cmj

  We've found a bug for you!
  /Users/cristianocalcagno/GitHub/rescript-compiler/jscomp/gentype_tests/typescript-react-example/src/V4.res

  Some record fields are undefined: y

FAILED: cannot make progress due to previous errors.
```

**Comment by mattdamon108** [comment]
> One more issue. An error without location is reported when a required prop is not supplied.
> 
> Example:
> 
> ```rescript
> module C = {
>   @react.component
>   let make = ( ~x, ~y) => React.string(x++y)
> }
> 
> let c = <C x="abc" />
> ```
> 
> Compilation error:
> 
> ```shell
> FAILED: src/V4.cmj
> 
>   We've found a bug for you!
>   /Users/cristianocalcagno/GitHub/rescript-compiler/jscomp/gentype_tests/typescript-react-example/src/V4.res
> 
>   Some record fields are undefined: y
> 
> FAILED: cannot make progress due to previous errors.
> ```

I'll look into it. Anyway I'm working on the raising errors with loc. It seems related issues.

**Comment by cristianoc** [comment]
It would be nice to have an explicit mode where after formatting, the result of running the ppx is left as normal code.
This could be very useful for debugging, and to make sure that the ppx transformation is not magic but is opt-in: the user can decide whether or not to use it. And might decide to use it only once, but check in the result directly.

**Comment by cristianoc** [comment]
> > One more issue. An error without location is reported when a required prop is not supplied.
> > Example:
> > ```rescript
> > module C = {
> >   @react.component
> >   let make = ( ~x, ~y) => React.string(x++y)
> > }
> > 
> > let c = <C x="abc" />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Compilation error:
> > ```shell
> > FAILED: src/V4.cmj
> > 
> >   We've found a bug for you!
> >   /Users/cristianocalcagno/GitHub/rescript-compiler/jscomp/gentype_tests/typescript-react-example/src/V4.res
> > 
> >   Some record fields are undefined: y
> > 
> > FAILED: cannot make progress due to previous errors.
> > ```
> 
> I'll look into it. Anyway I'm working on the raising errors with loc. It seems related issues.

I can confirmed it is a location issue: int he following:
```rescript
// let c1 = C.make({x: "abc"})
// let c2 = <C x="abc" />
```
only `c2` gives an error without location.

**Comment by mattdamon108** [comment]
11 shoud be Fixed https://github.com/rescript-lang/syntax/pull/620/commits/89561dfdb33c189351117835491e1af7eb28e98f
```
FAILED: src/AnyName.cmj

  We've found a bug for you!
  /Users/woonki/Github/projects/react-ppx-v4/src/AnyName.res:7:11-22

  5 │ 
  6 │ // let c1 = C.make({x: "abc"})
  7 │ let c2 = <C x="abc" />
  8 │ // let c3 = React.jsx(C.make, {x: "abc"})
  9 │ 

  Some record fields are undefined: y

FAILED: cannot make progress due to previous errors.
>>>> Finish compiling(exit: 1)
```

**Comment by cristianoc** [comment]
What's up with 8: is there an actual example that has a problem?

**Comment by mattdamon108** [comment]
It's mysterious 😄 You confirm that it is compiled just fine. The same example is not compiled in my local.
<img width="405" alt="image" src="https://user-images.githubusercontent.com/36963424/180508656-35ae3a30-c4d0-4b60-a662-160616e90ff8.png">

**Comment by cristianoc** [comment]
Let me try to add it to a PR then.

**Comment by mattdamon108** [comment]
Ah! I noticed it is different between V4 "classic" and "automatic"

**Comment by cristianoc** [comment]
Done: https://github.com/rescript-lang/rescript-compiler/pull/5576

**Comment by cristianoc** [comment]
Is there a bug in automatic?

**Comment by mattdamon108** [comment]
I don't think so. I think It is caused where the prop value is used. See the uppercase example here.
```rescript
let optionMap = (x, f) =>
  switch x {
  | None => None
  | Some(v) => Some(f(v))
  }

module A = {
  @react.component
  let make = (~name=?) => <div ?name />
}

module Parens = {
  @react.component
  let make = (~name=?) => {
    <A
      name=?{
        let _ = ()
        name->Option.map(x => x)
      }
    />
  }
}

module WithoutParens = {
  @react.component
  let make = (~name=?) => {
    <A name=?{name->optionMap(x => x)} />
  }
}
```
In the uppercase example, it is not compiled in both "classic" and "automatic". Because only V4 + "classic" for lowercase put the prop value in the labeled arg.
```rescript
React.createElementVariadic(
  "div"
  ~props=ReactDOMRe.domProps(~name=?{name->optionMap(x => x)}),
  ()
)
```

**Comment by mattdamon108** [comment]
* lowercase, "classic" -> labeled arg
* uppercase, "automatic" -> record
* lowercase, "classic" -> record
* uppercase, "automatic" -> record

**Comment by mattdamon108** [comment]
```rescript
React.createElement(A.make, {name: ?name->optionMap(x => x)})
```
This is not compiled.

**Comment by mattdamon108** [comment]
The printed output has difference between them
```rescript
// in labeled arg
?{name->optionMap(x => x)}

// in record
?name->optionMap(x => x)
```

**Comment by cristianoc** [comment]
We can remove `        let _ = ()` as it makes no difference and only creates confusion.

**Comment by cristianoc** [comment]
I think this difference is more relevant:
`optionMap(name, x => x)` vs `name->optionMap(x => x)`

**Comment by cristianoc** [comment]
```rescript
@@jsxConfig({version: 4})

let optionMap = (x, f) =>
  switch x {
  | None => None
  | Some(v) => Some(f(v))
  }

module A = {
  @react.component
  let make = (~name=?) => <div ?name />
}

module Direct = {
  @react.component
  let make = (~name=?) => {
    <A
      name=?{
      optionMap(name, x => x)
      }
    />
  }
}

module Pipe = {
  @react.component
  let make = (~name=?) => {
    <A
      name=?{
        name->optionMap(x => x)
      }
    />
  }
}
```

**Comment by cristianoc** [comment]
Looks like it's the implementation of `->` in the compiler creating trouble.
And it might also create trouble whenever it's used inside function application.

**Comment by cristianoc** [comment]
Here's a complete self-contained example without any ppx:
```rescript
let optionMap = (x, f) =>
  switch x {
  | None => None
  | Some(v) => Some(f(v))
  }

type props<'name> = {key?: string, name?: string}

let name = None

let ok = {name: ?optionMap(name, x => x)}
let bad = {name: ?name->optionMap(x => x)}
```

**Comment by cristianoc** [comment]
Compiler bug on `->`.

**Comment by cristianoc** [comment]
To confirm: this
```res
let ok = {name:  @foo (optionMap(name, x => x))}
let bad = {name: @foo (name->optionMap(x => x))}
```

formats to

```res
let ok = {name: @foo optionMap(name, x => x)}
let bad = {name: name->optionMap(x => x)}
```

which means the annotation `@foo` is eaten up by formatting.

**Comment by cristianoc** [comment]
And to simplify further, this:
```res
let ok =  @foo (optionMap(name, x => x))
let bad = @foo (name->optionMap(x => x))
```

formats to this:
```res
let ok = @foo optionMap(name, x => x)
let bad = name->optionMap(x => x)
```

**Comment by cristianoc** [comment]
Added examples here: https://github.com/rescript-lang/syntax/pull/621

**Comment by mattdamon108** [comment]
Wow! yes, I've mentioned the operator precedence is suspicious though https://github.com/rescript-lang/syntax/issues/615#issuecomment-1191083087 I think you're right and your approach will solve this issue!

**Comment by cristianoc** [comment]
Even though formatting is an issue, that needs to be fixed, I don't think it can be a cause of this (there's no printing involved, just parsing).
So I suspect there's at least one more issue after parsing, happening inside the compiler when `->` is processed (as parsing seems OK).

**Comment by cristianoc** [comment]
The error one gets is the same as one would get by turning `<A name=?{optionMap(name, x => x)} />` into `<A name={optionMap(name, x => x)} />` i.e. removing `?` i.e. removing `@ns.optional` from the AST.

**Comment by mattdamon108** [comment]
> The error one gets is the same as one would get by turning `<A name=?{optionMap(name, x => x)} />` into `<A name={optionMap(name, x => x)} />` i.e. removing `?` i.e. removing `@ns.optional` from the AST.

Not sure I understand what you meant. Do you mean that there is still an error in the example without`@ns.optional`?

**Comment by cristianoc** [comment]
As I said above there's no printing when you compiler your original example: just parsing followed by type checking.
So I suspect a second bug is causing that.

**Comment by cristianoc** [comment]
Here's the bug: https://github.com/rescript-lang/rescript-compiler/pull/5581

**Comment by cristianoc** [comment]
Merged.

**Comment by mattdamon108** [comment]
I’ll test it

**Comment by mattdamon108** [comment]
> As I said above there's no printing when you compile your original example: just parsing followed by type checking.
> So I suspect a second bug is causing that.

Now I get it, thanks!

**Comment by mattdamon108** [comment]
I can confirm it fixed! 👍

**Comment by cristianoc** [comment]
Congrats on completing the first version of PPX V4.
This is one of the most elaborate features ever introduced in ReScript.

**Comment by mattdamon108** [comment]
@cristianoc I reopen issue no. 8. I found it has the same error with 1 arity fn.

<img width="401" alt="image" src="https://user-images.githubusercontent.com/36963424/180591062-23fca7a9-e38a-4900-bb3a-c27d0685096f.png">

**Comment by mattdamon108** [comment]
I'm not sure but suspicious here. https://github.com/rescript-lang/rescript-compiler/blob/e7c408ee63b86327bde3783a68cf673d475b4c8f/jscomp/frontend/ast_exp_apply.ml#L180

**Comment by mattdamon108** [comment]
Ah! I think `Pexp_ident` case needs to be added. Can I make a PR to fix it?

**Comment by cristianoc** [comment]
> Ah! I think `Pexp_ident` case needs to be added. If I'm right, can I make a PR to fix it?

Would you like to try to fix it directly?

**Comment by mattdamon108** [comment]
> > Ah! I think `Pexp_ident` case needs to be added. If I'm right, can I make a PR to fix it?
> 
> Would you like to try to fix it directly?

Yes, I'll give it a try!

**Comment by cristianoc** [comment]
Start by adding an example, just like the previous PR.

**Comment by mattdamon108** [comment]
If you're not in hurry, can I start it when I get back?

**Comment by mattdamon108** [comment]
Issue no. 8 is fixed by https://github.com/rescript-lang/rescript-compiler/pull/5585

**Comment by mattdamon108** [comment]
> Congrats on completing the first version of PPX V4. This is one of the most elaborate features ever introduced in ReScript.

Thank you for everything. It can't be done without your help and lead. I learned a lot. I'll keep contributing to the compiler, tools, and community as much as I can.
